File: src/android/ShieldFraudPlugin.java
package com.shieldfraud;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.Nullable;
import com.shield.android.BlockedDialog;
import com.shield.android.Callback;
import com.shield.android.DeviceIntelligence;
import com.shield.android.Environment;
import com.shield.android.LogLevel;
import com.shield.android.Result;
import com.shield.android.Shield;
import com.shield.android.ShieldConfig;
import com.shield.android.ShieldCrossPlatformHelper;
import com.shield.android.ShieldCrossPlatformParams;
import com.shield.android.ShieldError;
import com.shield.android.ShieldFactory;
import com.shield.android.ShieldUserData;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class ShieldFraudPlugin extends CordovaPlugin {

    private static Shield shieldInstance;
    private static final Handler mainHandler = new Handler(Looper.getMainLooper());
    private static final DeviceResultCallbackRouter deviceResultCallbackRouter = new DeviceResultCallbackRouter();

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        try {
            if ("initShieldFraud".equals(action)) {
                initShieldFraud(callbackContext, args);
                return true;
            } else if ("getSessionID".equals(action)) {
                getSessionId(callbackContext);
                return true;
            } else if ("getDeviceResult".equals(action)) {
                getDeviceResult(callbackContext);
                return true;
            } else if ("sendAttributes".equals(action)) {
                sendAttributes(callbackContext, args);
                return true;
            } else if ("sendDeviceSignature".equals(action)) {
                sendDeviceSignature(callbackContext, args);
                return true;
            } else if ("isShieldInitialized".equals(action)) {
                isShieldInitialized(callbackContext);
                return true;
            }
        } catch (Exception exception) {
            if ("initShieldFraud".equals(action)) {
                shieldInstance = null;
                deviceResultCallbackRouter.clear();
            }
            sendExceptionError(callbackContext, action, exception);
            return true;
        }
        return false;
    }

    @Override
    public void onReset() {
        deviceResultCallbackRouter.clear();
        super.onReset();
    }

    private void initShieldFraud(CallbackContext callbackContext, JSONArray args) {
        if (args == null) {
            callbackContext.error("Invalid arguments");
            return;
        }

        JSONObject payload = args.optJSONObject(0);
        if (payload == null) {
            callbackContext.error("Invalid arguments");
            return;
        }

        String siteID = payload.optString("siteID", "").trim();
        String key = payload.optString("secretKey", "").trim();

        if (siteID.isEmpty() || key.isEmpty()) {
            callbackContext.error("siteID and secretKey are required");
            return;
        }

        boolean enableDeviceResultListener = payload.optBoolean("enableDeviceResultListener", false);
        if (ShieldFraudPlugin.shieldInstance != null) {
            if (enableDeviceResultListener) {
                deviceResultCallbackRouter.setCallbackContext(callbackContext);
                JSONObject cachedResult = shieldInstance.getLatestDeviceResult();
                if (cachedResult != null) {
                    deviceResultCallbackRouter.replayCachedResult(cachedResult);
                }
            } else {
                runOnMainThread(() -> callbackContext.success(1));
            }
            return;
        }

        ShieldConfig shieldConfig = new ShieldConfig(siteID, key);
        shieldConfig.setEnvironment(parseEnvironment(payload.optInt("environment", 0)));
        shieldConfig.setLogLevel(parseLogLevel(payload.optInt("logLevel", 0)));
        shieldConfig.setBlockScreenRecording(payload.optBoolean("blockScreenRecording", false));
        String partnerId = payload.optString("partnerId", "");
        if (!partnerId.isEmpty()) {
            shieldConfig.setPartnerId(partnerId);
        }
        Object needBackgroundListenerValue = payload.opt("needBackgroundListener");
        if (needBackgroundListenerValue instanceof Boolean) {
            boolean needBackgroundListener = (Boolean) needBackgroundListenerValue;
            shieldConfig.setNeedBackgroundListener(needBackgroundListener);
        }

        String crossPlatformName = payload.optString("crossPlatformName", "").trim();
        String crossPlatformVersion = payload.optString("crossPlatformVersion", "").trim();
        if (!crossPlatformName.isEmpty() && !crossPlatformVersion.isEmpty()) {
            ShieldCrossPlatformHelper.setCrossPlatformParameters(
                    new ShieldCrossPlatformParams(crossPlatformName, crossPlatformVersion)
            );
        }

        JSONObject dialogArg = payload.optJSONObject("blockedDialog");
        if (dialogArg != null) {
            String dialogTitle = dialogArg.optString("title", "");
            String dialogBody = dialogArg.optString("body", "");
            shieldConfig.setBlockedDialog(new BlockedDialog(dialogTitle, dialogBody));
        }

        Application application = cordova.getActivity().getApplication();
        if (enableDeviceResultListener) {
            deviceResultCallbackRouter.setCallbackContext(callbackContext);
            ShieldFraudPlugin.shieldInstance = ShieldFactory.createShieldWithCallback(
                    application,
                    shieldConfig,
                    deviceResultCallbackRouter
            );
            return;
        }

        ShieldFraudPlugin.shieldInstance = ShieldFactory.createShield(application, shieldConfig);
        runOnMainThread(() -> callbackContext.success(1));
    }

    private void isShieldInitialized(CallbackContext callbackContext) {
        runOnMainThread(() -> callbackContext.success(ShieldFraudPlugin.shieldInstance != null ? 1 : 0));
    }

    private void getSessionId(CallbackContext callbackContext) {
        Shield shield = requireShield(callbackContext);
        if (shield == null) {
            return;
        }

        runOnMainThread(() -> callbackContext.success(shield.getSessionId()));
    }

    private void getDeviceResult(CallbackContext callbackContext) {
        Shield shield = requireShield(callbackContext);
        if (shield == null) {
            return;
        }

        JSONObject result = shield.getLatestDeviceResult();
        if (result != null) {
            runOnMainThread(() -> callbackContext.success(result));
        } else {
            runOnMainThread(() -> callbackContext.error("No device result available"));
        }
    }

    private void sendAttributes(CallbackContext callbackContext, JSONArray args) {
        Shield shield = requireShield(callbackContext);
        if (shield == null) {
            return;
        }

        if (args == null) {
            callbackContext.error("Invalid arguments");
            return;
        }

        try {
            String screenName = args.optString(0);
            JSONObject object = args.getJSONObject(1);
            Map<String, String> data = jsonObjectToHashMap(object);

            shield.sendAttributesWithCallback(screenName, data, new Callback<String>() {
                @Override
                public void onCallback(Result<String> result) {
                    handleStringResult(result, callbackContext);
                }
            });
        } catch (JSONException e) {
            sendExceptionError(callbackContext, "sendAttributes", e);
        }
    }
    
    private void sendDeviceSignature(CallbackContext callbackContext, JSONArray args) {
        Shield shield = requireShield(callbackContext);
        if (shield == null) {
            return;
        }
    
        if (args == null || args.length() == 0) {
            callbackContext.error("Invalid arguments");
            return;
        }
    
        JSONObject payload = args.optJSONObject(0);
        if (payload == null) {
            callbackContext.error("Invalid arguments");
            return;
        }
    
        Object screenNameValue = payload.opt("screenName");
        if (!(screenNameValue instanceof String)) {
            callbackContext.error("Invalid arguments");
            return;
        }
    
        String screenName = (String) screenNameValue;
        String userId = null;
    
        if (payload.has("userId") && !payload.isNull("userId")) {
            Object userIdValue = payload.opt("userId");
            if (!(userIdValue instanceof String)) {
                callbackContext.error("Invalid arguments");
                return;
            }
    
            userId = (String) userIdValue;
        }
    
        ShieldUserData userData = new ShieldUserData(screenName);
    
        if (userId != null && !userId.isEmpty()) {
            userData.setUserId(userId);
        }
    
        shield.sendDeviceSignatureWithCallback(userData, new Callback<String>() {
            @Override
            public void onCallback(Result<String> result) {
                handleStringResult(result, callbackContext);
            }
        });
    }

    private Shield requireShield(CallbackContext callbackContext) {
        if (ShieldFraudPlugin.shieldInstance == null) {
            callbackContext.error("Initialize SDK before calling this method");
            return null;
        }

        return ShieldFraudPlugin.shieldInstance;
    }

    private Environment parseEnvironment(int envValue) {
        switch (envValue) {
            case 1:
                return Environment.DEV;
            case 2:
                return Environment.STAGING;
            default:
                return Environment.PROD;
        }
    }

    private LogLevel parseLogLevel(int logValue) {
        switch (logValue) {
            case 1:
                return LogLevel.INFO;
            case 2:
                return LogLevel.DEBUG;
            case 3:
                return LogLevel.VERBOSE;
            default:
                return LogLevel.NONE;
        }
    }

    private static String shieldErrorToMessage(@Nullable ShieldError shieldError) {
        if (shieldError == null) {
            return "Unknown error";
        }

        if (shieldError.getErrorMessage() != null && !shieldError.getErrorMessage().isEmpty()) {
            return shieldError.getErrorMessage();
        }

        if (shieldError.getErrorCode() != null && !shieldError.getErrorCode().isEmpty()) {
            return shieldError.getErrorCode();
        }

        return shieldError.toString();
    }

    private static void runOnMainThread(Runnable runnable) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            runnable.run();
            return;
        }

        mainHandler.post(runnable);
    }

    private static void handleStringResult(Result<String> result, CallbackContext callbackContext) {
        runOnMainThread(() -> {
            try {
                if (result instanceof Result.Success) {
                    String sessionId = ((Result.Success<String>) result).getData();
                    callbackContext.success(sessionId != null ? sessionId : "");
                } else if (result instanceof Result.Failure) {
                    ShieldError shieldError = ((Result.Failure<String>) result).getError();
                    callbackContext.error(shieldErrorToMessage(shieldError));
                } else {
                    callbackContext.error("Unexpected SDK result");
                }
            } catch (Exception exception) {
                sendExceptionError(callbackContext, "processing SDK callback", exception);
            }
        });
    }

    private static void sendExceptionError(
            CallbackContext callbackContext,
            String action,
            Exception exception
    ) {
        String detail = exception.getMessage();
        if (detail == null || detail.isEmpty()) {
            detail = exception.getClass().getSimpleName();
        }
        String message = "Failed to " + action + ": " + detail;
        runOnMainThread(() -> callbackContext.error(message));
    }

    private static final class DeviceResultCallbackRouter implements Callback<DeviceIntelligence> {
        private volatile CallbackContext callbackContext;

        void setCallbackContext(CallbackContext callbackContext) {
            this.callbackContext = callbackContext;
        }

        void clear() {
            callbackContext = null;
        }

        void replayCachedResult(JSONObject cachedResult) {
            sendSuccess(cachedResult);
        }

        @Override
        public void onCallback(Result<DeviceIntelligence> result) {
            runOnMainThread(() -> {
                CallbackContext currentContext = callbackContext;
                if (currentContext == null) {
                    return;
                }

                try {
                    if (result instanceof Result.Success) {
                        DeviceIntelligence intelligence = ((Result.Success<DeviceIntelligence>) result).getData();
                        JSONObject payload = intelligence != null ? intelligence.getData() : null;
                        sendSuccess(payload != null ? payload : new JSONObject());
                    } else if (result instanceof Result.Failure) {
                        ShieldError shieldError = ((Result.Failure<DeviceIntelligence>) result).getError();
                        sendError(shieldErrorToMessage(shieldError));
                    } else {
                        sendError("Unexpected SDK result");
                    }
                } catch (Exception exception) {
                    String detail = exception.getMessage();
                    sendError(detail != null && !detail.isEmpty()
                            ? detail
                            : exception.getClass().getSimpleName());
                }
            });
        }

        private void sendSuccess(JSONObject payload) {
            runOnMainThread(() -> {
                CallbackContext currentContext = callbackContext;
                if (currentContext == null) {
                    return;
                }
                PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, payload);
                pluginResult.setKeepCallback(true);
                currentContext.sendPluginResult(pluginResult);
            });
        }

        private void sendError(String message) {
            runOnMainThread(() -> {
                CallbackContext currentContext = callbackContext;
                if (currentContext == null) {
                    return;
                }
                PluginResult pluginResult = new PluginResult(PluginResult.Status.ERROR, message);
                pluginResult.setKeepCallback(true);
                currentContext.sendPluginResult(pluginResult);
            });
        }
    }

    private static HashMap<String, String> jsonObjectToHashMap(JSONObject jsonObject) throws JSONException {
        HashMap<String, String> hashMap = new HashMap<>();
        Iterator<String> iterator = jsonObject.keys();
        while (iterator.hasNext()) {
            String key = iterator.next();
            String value = jsonObject.getString(key);
            hashMap.put(key, value);
        }
        return hashMap;
    }
}

File: src/ios/ShieldFraudPlugin.swift
import ShieldFraud

@objc(ShieldFraudPlugin) class ShieldFraudPlugin : CDVPlugin {

    private static var shieldInstance: Shield?

    private func sendPluginResult(_ pluginResult: CDVPluginResult?, callbackId: String) {
        guard let pluginResult = pluginResult else {
            return
        }

        DispatchQueue.main.async {
            self.commandDelegate.send(pluginResult, callbackId: callbackId)
        }
    }

    private func requireShield(command: CDVInvokedUrlCommand, methodName: String) -> Shield? {
        guard let shield = ShieldFraudPlugin.shieldInstance else {
            let pluginResult = CDVPluginResult(
                status: .error,
                messageAs: "Initialize SDK before calling \(methodName)"
            )
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            return nil
        }
        return shield
    }

    private func shieldErrorMessage(_ error: ShieldError) -> String {
        if !error.errorMessage.isEmpty {
            return error.errorMessage
        }
        if !error.errorCode.isEmpty {
            return error.errorCode
        }
        return error.localizedDescription
    }

    private func serializeDeviceResult(_ data: [String: Any]) throws -> String {
        let jsonData = try JSONSerialization.data(withJSONObject: data, options: [])
        guard let dataString = String(data: jsonData, encoding: .utf8) else {
            throw NSError(
                domain: "ShieldFraudPlugin",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Device result is not valid UTF-8"]
            )
        }
        return dataString
    }

    private func serializationErrorMessage(_ error: Error) -> String {
        return "Failed to serialize device result: \(error.localizedDescription)"
    }

    private func registerDeviceResultListener(shield: Shield, callbackId: String) {
        shield.onDeviceResult { [weak self] intelligence, error in
            guard let self = self else {
                return
            }

            let pluginResult: CDVPluginResult?
            if let error = error {
                pluginResult = CDVPluginResult(
                    status: .error,
                    messageAs: self.shieldErrorMessage(error)
                )
            } else if let data = intelligence?.data {
                do {
                    let dataString = try self.serializeDeviceResult(data)
                    pluginResult = CDVPluginResult(status: .ok, messageAs: dataString)
                } catch {
                    pluginResult = CDVPluginResult(
                        status: .error,
                        messageAs: self.serializationErrorMessage(error)
                    )
                }
            } else {
                pluginResult = CDVPluginResult(
                    status: .error,
                    messageAs: "Device result data is unavailable"
                )
            }

            pluginResult?.setKeepCallbackAs(true)
            self.sendPluginResult(pluginResult, callbackId: callbackId)
        }
    }

    @objc(initShieldFraud:) func initShieldFraud(command: CDVInvokedUrlCommand) {
        guard command.arguments.count > 0,
              let payload = command.arguments[0] as? [String: Any],
              let siteID = payload["siteID"] as? String,
              let key = payload["secretKey"] as? String,
              !siteID.isEmpty,
              !key.isEmpty else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "siteID and secretKey are required")
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            return
        }

        let enableDeviceResultListener = payload["enableDeviceResultListener"] as? Bool ?? false
        if let shield = ShieldFraudPlugin.shieldInstance {
            if enableDeviceResultListener {
                registerDeviceResultListener(shield: shield, callbackId: command.callbackId)
            } else {
                let pluginResult = CDVPluginResult(status: .ok, messageAs: true)
                self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            }
            return
        }

        let rawEnv      = payload["environment"] as? Int ?? 0
        let environment = Environment(rawValue: rawEnv) ?? .prod

        // JS LogLevel: NONE=0, INFO=1, DEBUG=2, VERBOSE=3
        // iOS LogLevel: none=1, info=2, debug=3 (no VERBOSE — map to debug)
        let rawLog = payload["logLevel"] as? Int ?? 0
        let logLevel: LogLevel
        switch rawLog {
        case 1:  logLevel = .info
        case 2:  logLevel = .debug
        case 3:  logLevel = .debug  // VERBOSE not available on iOS, use debug
        default: logLevel = .none
        }

        if let crossPlatformName = payload["crossPlatformName"] as? String,
           let crossPlatformVersion = payload["crossPlatformVersion"] as? String,
           !crossPlatformName.isEmpty,
           !crossPlatformVersion.isEmpty {
            let params = ShieldCrossPlatformParams(name: crossPlatformName, version: crossPlatformVersion)
            ShieldCrossPlatformHelper.setCrossPlatformParameters(params)
        }

        let config = ShieldConfig(siteId: siteID, secretKey: key)
        config.environment = environment
        config.logLevel = logLevel
        if let partnerId = payload["partnerId"] as? String,
           !partnerId.isEmpty {
            config.partnerId = partnerId
        }
        if let dialogArg = payload["blockedDialog"] as? [String: String],
           let title = dialogArg["title"],
           let body = dialogArg["body"] {
            config.defaultBlockedDialog = BlockedDialog(title: title, body: body)
        }

        let shield = ShieldFactory.createShield(config: config)
        ShieldFraudPlugin.shieldInstance = shield

        if enableDeviceResultListener {
            registerDeviceResultListener(shield: shield, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: .ok, messageAs: true)
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(getSessionID:) func getSessionID(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run {
            guard let shield = self.requireShield(command: command, methodName: "getSessionID") else {
                return
            }
            let sessionId = shield.sessionId
            let pluginResult = CDVPluginResult(status: .ok, messageAs: sessionId)
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(getDeviceResult:) func getDeviceResult(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run {
            guard let shield = self.requireShield(command: command, methodName: "getDeviceResult") else {
                return
            }

            guard let deviceResult = shield.getLatestDeviceResult()?.data else {
                let pluginResult = CDVPluginResult(status: .error, messageAs: "No device result available")
                self.sendPluginResult(pluginResult, callbackId: command.callbackId)
                return
            }

            let pluginResult: CDVPluginResult?
            do {
                let dataString = try self.serializeDeviceResult(deviceResult)
                pluginResult = CDVPluginResult(status: .ok, messageAs: dataString)
            } catch {
                pluginResult = CDVPluginResult(
                    status: .error,
                    messageAs: self.serializationErrorMessage(error)
                )
            }
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(sendAttributes:) func sendAttributes(command: CDVInvokedUrlCommand) {
        guard command.arguments.count >= 2,
              let screenName = command.arguments[0] as? String,
              let data       = command.arguments[1] as? [String: String] else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "screenName and data are required")
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            return
        }
        self.commandDelegate.run {
            guard let shield = self.requireShield(command: command, methodName: "sendAttributes") else {
                return
            }

            shield.sendAttributes(screenName: screenName, data: data) { sessionId, error in
                let pluginResult: CDVPluginResult?
                if let error = error {
                    pluginResult = CDVPluginResult(
                        status: .error,
                        messageAs: self.shieldErrorMessage(error)
                    )
                } else {
                    pluginResult = CDVPluginResult(
                        status: .ok,
                        messageAs: sessionId ?? shield.sessionId
                    )
                }
                self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            }
        }
    }

    @objc(sendDeviceSignature:) func sendDeviceSignature(command: CDVInvokedUrlCommand) {
        guard command.arguments.count > 0,
            let payload = command.arguments[0] as? [String: Any] else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "Invalid arguments")
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            return
        }

        guard let screenName = payload["screenName"] as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "Invalid arguments")
            self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            return
        }

        var userId: String? = nil

        if let userIdValue = payload["userId"],
        !(userIdValue is NSNull) {
            guard let validUserId = userIdValue as? String else {
                let pluginResult = CDVPluginResult(status: .error, messageAs: "Invalid arguments")
                self.sendPluginResult(pluginResult, callbackId: command.callbackId)
                return
            }

            userId = validUserId
        }

        self.commandDelegate.run {
            guard let shield = self.requireShield(command: command, methodName: "sendDeviceSignature") else {
                return
            }

            let userData: ShieldUserData

            if let userId = userId,
               !userId.isEmpty {
                userData = ShieldUserData(
                    screenName: screenName,
                    userId: userId
                )
            } else {
                userData = ShieldUserData(
                    screenName: screenName
                )
            }

            shield.sendDeviceSignature(userData: userData) { sessionId, error in
                let pluginResult: CDVPluginResult?
                if let error = error {
                    pluginResult = CDVPluginResult(
                        status: .error,
                        messageAs: self.shieldErrorMessage(error)
                    )
                } else {
                    pluginResult = CDVPluginResult(
                        status: .ok,
                        messageAs: sessionId ?? shield.sessionId
                    )
                }
                self.sendPluginResult(pluginResult, callbackId: command.callbackId)
            }
        }
    }

    @objc(isShieldInitialized:) func isShieldInitialized(command: CDVInvokedUrlCommand) {
        let pluginResult = CDVPluginResult(
            status: .ok,
            messageAs: ShieldFraudPlugin.shieldInstance != nil
        )
        self.sendPluginResult(pluginResult, callbackId: command.callbackId)
    }
}
